src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs

```diff
@@ -121,99 +338,160 @@ private string CreateVaryByCacheKey(CachedVaryBy varyBy)
 
             if (cacheEntry is CachedResponse)
             {
-                // TODO: Compare cached request headers
-
-                // TODO: Content negotiation if there are multiple cached response formats?
-
-                // TODO: Verify content freshness, or else re-validate the data?
-
                 var cachedResponse = cacheEntry as CachedResponse;
+                var cachedResponseHeaders = new ResponseHeaders(cachedResponse.Headers);
 
-                var response = HttpContext.Response;
-                // Copy the cached status code and response headers
-                response.StatusCode = cachedResponse.StatusCode;
-                foreach (var pair in cachedResponse.Headers)
+                _responseTime = DateTimeOffset.UtcNow;
+                var age = _responseTime - cachedResponse.Created;
+                age = age > TimeSpan.Zero ? age : TimeSpan.Zero;
+
+                if (EntryIsFresh(cachedResponseHeaders, age, verifyAgainstRequest: true))
                 {
-                    response.Headers[pair.Key] = pair.Value;
-                }
+                    var response = HttpContext.Response;
+                    // Copy the cached status code and response headers
+                    response.StatusCode = cachedResponse.StatusCode;
+                    foreach (var header in cachedResponse.Headers)
+                    {
+                        response.Headers.Add(header);
+                    }
 
-                // TODO: Update cache headers (Age)
-                response.Headers["Served_From_Cache"] = DateTime.Now.ToString();
+                    response.Headers[HeaderNames.Age] = ((int)Math.Min(age.TotalSeconds, int.MaxValue)).ToString(CultureInfo.InvariantCulture);
```

Comment:
We could. RFC says minimum of 31 bits of non-negative integer range so `int` is enough but that's a minimum requirement so I guess we can also use long.